### PR TITLE
[SV Choice] Swedish support in Choice recognizer

### PR DESCRIPTION
**Pull request by pete1854**
This is the draft PR suggesting to add sv-SE support for the Choice recognizer.
Part of #1493.

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj [hunk 1]
```diff
@@ -326,6 +336,8 @@
     <None Update="English\DateTimeDefinitions.tt">
     </None>
     <None Update="English\NumbersDefinitions.tt">
+      <Generator>TextTemplatingFileGenerator</Generator>
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: Why this change?

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj [hunk 2]
```diff
@@ -379,12 +391,27 @@
     </None>
     <None Update="Spanish\NumbersWithUnitDefinitions.tt">
     </None>
+    <None Update="Swedish\ChoiceDefinitions.tt">
+      <Generator>TextTemplatingFileGenerator</Generator>
+    </None>
+    <None Update="Swedish\NumbersDefinitions.tt">
+      <Generator>TextTemplatingFileGenerator</Generator>
+      <CustomToolNamespace></CustomToolNamespace>
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: We should be careful to follow exactly the same attributes are the structure for other cultures. For some reason it seems the new entries are a little inconsistent.

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj [hunk 3]
```diff
@@ -379,12 +391,27 @@
...
+      <LastGenOutput>NumbersWithUnitDefinitions.cs</LastGenOutput>
+    </None>
   </ItemGroup>
 
   <ItemGroup>
     <DotNetCliToolReference Include="T5.TextTransform.Tool" Version="1.1.0-*" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: What is this service?

File: .NET/Microsoft.Recognizers.Definitions.Common/Swedish/DateTimeDefinitions.cs [hunk 4]
```diff
@@ -0,0 +1,753 @@
+﻿//------------------------------------------------------------------------------
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions.Common/Swedish/DateTimeDefinitions.cs]: It would be best to keep only the Choice-related changes in this PR.

File: .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj [hunk 5]
```diff
@@ -79,6 +79,9 @@
     <Compile Include="..\Microsoft.Recognizers.Definitions.Common\Spanish\DateTimeDefinitions.cs" Link="Spanish\DateTimeDefinitions.cs" />
     <Compile Include="..\Microsoft.Recognizers.Definitions.Common\Spanish\NumbersDefinitions.cs" Link="Spanish\NumbersDefinitions.cs" />
     <Compile Include="..\Microsoft.Recognizers.Definitions.Common\Spanish\NumbersWithUnitDefinitions.cs" Link="Spanish\NumbersWithUnitDefinitions.cs" />
+    <Compile Include="..\Microsoft.Recognizers.Definitions.Common\Swedish\ChoiceDefinitions.cs" Link="Swedish\ChoiceDefinitions.cs" />
+    <Compile Include="..\Microsoft.Recognizers.Definitions.Common\Swedish\NumbersDefinitions.cs" Link="Swedish\NumbersDefinitions.cs" />
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj]: Let's include only Choice for now ,ok?

File: .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj [hunk 6]
```diff
@@ -104,5 +108,9 @@
       <IncludeAssets>runtime; build; native; contentfiles; analyzers</IncludeAssets>
     </PackageReference>
   </ItemGroup>
+  
+  <ItemGroup>
+    <Service Include="{508349b6-6b84-4df5-91f0-309beebad82d}" />
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj]: Same as above?

File: .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs [hunk 7]
```diff
@@ -60,7 +60,7 @@ public List<ModelResult> Parse(string query)
                 // Only support "subtype" for English for now
                 // As some languages like German, we miss handling some subtypes between "decimal" and "integer"
                 if (!string.IsNullOrEmpty(o.Type) &&
-                    Constants.ValidSubTypes.Contains(o.Type) && extractorType.Contains(Constants.ENGLISH))
+                    Constants.ValidSubTypes.Contains(o.Type) && (extractorType.Contains(Constants.ENGLISH) | extractorType.Contains(Constants.SWEDISH)))
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs]: It would be best to declare a method to check for supported cultures, instead of just adding new ``ìf``` conditions.

File: Specs/Choice/Swedish/BooleanModel.json [hunk 8]
```diff
@@ -0,0 +1,506 @@
+[
+  {
+    "Input": "sant",
+    "NotSupportedByDesign": "python,java",
```
Comment (tellarin) [inline Specs/Choice/Swedish/BooleanModel.json]: As currently only Swedish is implemented, you can add a separate "NotSupported" tag for "javascript". The build will fail as JS support doesn't exist yet.
"NotSupportted" indicates that support will come later. "NotSupportedByDesign" indicates that either it won't be implemented, or that there are no current plans to do so.

File: .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
+﻿using Microsoft.Recognizers.Text.DataDrivenTests;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+
+namespace Microsoft.Recognizers.Text.Number.Tests
+{
+    [TestClass]
+    public class TestNumber_Swedish : TestBase
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs]: Let's remove these Number test skeleton classes from this PR. Build is failing as they don't currently run.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs]:
```
Failed   NumberModel
Error Message:
 Class Initialization method Microsoft.Recognizers.Text.Number.Tests.TestNumber_Swedish.ClassInitialize threw exception. System.IO.DirectoryNotFoundException: System.IO.DirectoryNotFoundException: Could not find a part of the path 'C:\projects\recognizers-text\Specs\Number\Swedish'..
Stack Trace:
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileSystemEnumerableIterator`1.CommonInit()
   at System.IO.FileSystemEnumerableIterator`1..ctor(String path, String originalUserPath, String searchPattern, SearchOption searchOption, SearchResultHandler`1 resultHandler, Boolean checkHost)
   at System.IO.Directory.GetFiles(String path, String searchPattern)
   at Microsoft.Recognizers.Text.DataDrivenTests.TestResourcesExtensions.InitFromTestContext(TestResources resources, TestContext context) in C:\projects\recognizers-text\.NET\Microsoft.Recognizers.Text.DataDrivenTests\TestResourcesExtensions.cs:line 18
   at Microsoft.Recognizers.Text.Number.Tests.TestNumber_Swedish.ClassInitialize(TestContext context) in C:\projects\recognizers-text\.NET\Microsoft.Recognizers.Text.DataDrivenTests\Number\TestNumber_Swedish.cs:line 14
```

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj (hunk 1 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: Sorry - shouldn't be there. My bad.

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj (hunk 3 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: When adding at .tt file VS adds this reference by default.

File: .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj (hunk 5 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj]: Yes. I actually had only one branch containing both recognizers. I divided it up and these are remnants from that branch. Sorry.

File: .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj (hunk 6 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj]: yes

File: .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs (hunk 7 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs]: Yes probably. Or, maybe move the responsability for this to the extractor e.g: `extractorType.Supports("subtype")`, or something

File: Specs/Choice/Swedish/BooleanModel.json (hunk 8 above)
Comment (pete1854) [reply Specs/Choice/Swedish/BooleanModel.json]: From my part I'll not add support for python, java or even JS in the forseeable future. Should I change them all to "NotSupported" in case someone else feels inclined to contribute? Or maybe put JS to NotSupported and the rest as NotSupportedByDesign? I don't know.

File: .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs (hunk 9 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs]: ok

File: Specs/Choice/Swedish/BooleanModel.json (hunk 8 above)
Comment (pete1854) [reply Specs/Choice/Swedish/BooleanModel.json]: I'll change as per your suggestion.

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj (hunk 1 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: Fixed

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj (hunk 2 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: Fixed

File: .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj (hunk 3 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Microsoft.Recognizers.Definitions.Common.csproj]: Fixed (removed)

File: .NET/Microsoft.Recognizers.Definitions.Common/Swedish/DateTimeDefinitions.cs (hunk 4 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions.Common/Swedish/DateTimeDefinitions.cs]: Fixed

File: .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj (hunk 6 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Definitions/Microsoft.Recognizers.Definitions.csproj]: Fixed

File: .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs (hunk 9 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Text.DataDrivenTests/Number/TestNumber_Swedish.cs]: Fixed

File: .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs (hunk 7 above)
Comment (pete1854) [reply .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs]: @tellarin what do you recommend; should I try and add the function in the  extractor, or, should the function be implemented directly in the AbstractNumberModel class? I'm not that sure of your design philosophy yet, so, your advice is welcome.

Comment (pete1854) [reply .NET/Microsoft.Recognizers.Text.Number/Models/AbstractNumberModel.cs]: I'll remove this change from this PR since it's not relevant in this context.

**Comment by tellarin** [comment]
@pete1854 Merged!
Quick comment though, for next PRs, please write the commit comments in English, ok? Thanks.
